Quota

yadvr opened the pull request:
WIP for single review

bhaisaab commented:
@abhinandanprateek too many changes, you need to fix the source/target branches?

abhinandanprateek committed:
quota: initial commit

abhinandanprateek committed:
adding quota API to commands.properties

abhinandanprateek committed:
updating sql scripts

abhinandanprateek committed:
adding quota credit api

abhinandanprateek committed:
merging fixes

agneya2001 committed:
raised the version to 4.5.2 in pom

abhinandanprateek committed:
CLOUDSTACK-8592: add quota email template management commands

yadvr committed:
schema-451to452: Re-add SAML SQL changes reverted in 5ea1a89

Signed-off-by: Rohit Yadav <rohit.yadav@shapeblue.com>

yadvr committed:
apidoc: add section for quota APIs

Signed-off-by: Rohit Yadav <rohit.yadav@shapeblue.com>

yadvr committed:
quota: fix db bean issue

Signed-off-by: Rohit Yadav <rohit.yadav@shapeblue.com>

agneya2001 committed:
CLOUDSTACK-8592: Adding email management APIs

agneya2001 committed:
CLOUDSTACK-8592: updated quota email templates

abhinandanprateek committed:
adding pom dependencies for usage

abhinandanprateek committed:
CLOUDSTACK-8592: adding quota usaage and balance tables

abhinandanprateek committed:
CLOUDSTACK-8592: creating respective VO objects

abhinandanprateek committed:
CLOUDSTACK-8592: adding corresponding DAOs with stubs

abhinandanprateek committed:
CLOUDSTACK-8592: adding corresponding DAOs with stubs

yadvr committed:
ui: Allow plugins to be shown on navigation bar, if they want to

Signed-off-by: Rohit Yadav <rohit.yadav@shapeblue.com>

yadvr committed:
ui/plugins: Add initial Quota plugin

- Adds a stub
- Adds a Quota tab on the user's page

Signed-off-by: Rohit Yadav <rohit.yadav@shapeblue.com>

yadvr committed:
ui/scripts/cloudStack: simplify plugin loop and display check

Signed-off-by: Rohit Yadav <rohit.yadav@shapeblue.com>

abhinandanprateek committed:
CLOUDSTACK-8592: Adding new APIs and implementing the API stubs

abhinandanprateek committed:
CLOUDSTACK-8592: keep track of quota job runs - adding schema and daos

abhinandanprateek committed:
CLOUDSTACK-8592: adding QuotaEditResourceMappingCmd

abhinandanprateek committed:
CLOUDSTACK-8592: refactoring, cleanup

abhinandanprateek committed:
CLOUDSTACK-8592: calculation of quota

abhinandanprateek committed:
CLOUDSTACK-8592: reverting debug statement

abhinandanprateek committed:
CLOUDSTACK-8592: Schema updates

abhinandanprateek committed:
CLOUDSTACK-8592: first cut implementation of quota statement

yadvr committed:
setup/db: fix saml sql files, and create quota tables only if that don't exist

Signed-off-by: Rohit Yadav <rohit.yadav@shapeblue.com>

yadvr committed:
setup/db: fix schema issues, put quota stuff in cloud_usage db

Add missing column to cloud_usage

Signed-off-by: Rohit Yadav <rohit.yadav@shapeblue.com>

yadvr committed:
setup/db: fix schema issues, put quota stuff in cloud_usage db

Signed-off-by: Rohit Yadav <rohit.yadav@shapeblue.com>

yadvr committed:
quota: fix quotaRefresh API response keys

Signed-off-by: Rohit Yadav <rohit.yadav@shapeblue.com>

yadvr committed:
quota: fix quota refresh API, reuse success response

Signed-off-by: Rohit Yadav <rohit.yadav@shapeblue.com>

yadvr committed:
quota: rename quotamapping to quotamappingcmd

Signed-off-by: Rohit Yadav <rohit.yadav@shapeblue.com>

yadvr committed:
client: fix label and usage

Signed-off-by: Rohit Yadav <rohit.yadav@shapeblue.com>

yadvr committed:
quota: use uniform response names

Signed-off-by: Rohit Yadav <rohit.yadav@shapeblue.com>

yadvr committed:
quota: take value as string in add credit cmd

API was not working, using strign param we can create a BigDecimal value

Signed-off-by: Rohit Yadav <rohit.yadav@shapeblue.com>

abhinandanprateek committed:
iCLOUDSTACK-8592: adding accountid and domainid to quota statement

yadvr committed:
quota: for BigDecimal use String in API cmd and responses

Signed-off-by: Rohit Yadav <rohit.yadav@shapeblue.com>

yadvr committed:
quota: inject quota balance dao to quota db utils to be used later

Signed-off-by: Rohit Yadav <rohit.yadav@shapeblue.com>

abhinandanprateek committed:
CLOUDSTACK-8592: added quota balance command that aggregates quota line items and balance information

yadvr committed:
quota-ui: fix settings view

Signed-off-by: Rohit Yadav <rohit.yadav@shapeblue.com>

abhinandanprateek committed:
CLOUDSTACK-8592: close transactions used for swicthing DBs

abhinandanprateek committed:
CLOUDSTACK-8592: formatting

yadvr committed:
quota: remove unwanted changes

Signed-off-by: Rohit Yadav <rohit.yadav@shapeblue.com>

abhinandanprateek committed:
CLOUDSTACK-8592: put a limit on number of usage records read for processing

abhinandanprateek committed:
iCLOUDSTACK-8592: Incremental update for credits and balance

yadvr committed:
quota: refactor and fix code around quota tariff

quota mapping is now called quota tariff
fixed db usage

Signed-off-by: Rohit Yadav <rohit.yadav@shapeblue.com>

yadvr committed:
quota: rename/refactor quota mapping usage to quotaTariff

Signed-off-by: Rohit Yadav <rohit.yadav@shapeblue.com>

abhinandanprateek committed:
CLOUDSTACK-8592: calculating balance, adjusting credits and API commands for getting quota statement and balances

yadvr committed:
quota: UI view for quota tariff

listing and editing supported; global settings removed

Signed-off-by: Rohit Yadav <rohit.yadav@shapeblue.com>

abhinandanprateek committed:
CLOUDSTACK-8592: increase precision

abhinandanprateek committed:
CLOUDSTACK-8592: in responses print values rounded to two decimal places

abhinandanprateek committed:
CLOUDSTACK-8592: added current balance and balance on a date

yadvr committed:
quota: fix quota statement api, db schema and implement UI tab for statement generation

Signed-off-by: Rohit Yadav <rohit.yadav@shapeblue.com>

abhinandanprateek committed:
CLOUDSTACK-8592: splitting out quotamanager into quotamanager and quotaservice

yadvr committed:
quota: add credit management in the ui

Signed-off-by: Rohit Yadav <rohit.yadav@shapeblue.com>

yadvr committed:
quota: add quota currency on tarriff table

Signed-off-by: Rohit Yadav <rohit.yadav@shapeblue.com>

yadvr committed:
quota: remove console.log statement

Signed-off-by: Rohit Yadav <rohit.yadav@shapeblue.com>

yadvr committed:
quota: add quota balance/credit statement in the statement tab

Signed-off-by: Rohit Yadav <rohit.yadav@shapeblue.com>

abhinandanprateek committed:
quota: added checks for boundary cases

yadvr committed:
quota: fix quota credits api, NPE

Signed-off-by: Rohit Yadav <rohit.yadav@shapeblue.com>

abhinandanprateek committed:
quota: fixed transaction, boundary conditions

yadvr committed:
quota: remove invalid ui code, add additional checks

Signed-off-by: Rohit Yadav <rohit.yadav@shapeblue.com>

yadvr committed:
quota: re-order date vs domain/account checking

Signed-off-by: Rohit Yadav <rohit.yadav@shapeblue.com>

abhinandanprateek committed:
quota: for quota calc from start date to end date, adjust the credits/debits done during this period

yadvr committed:
quota: add apache commons 3.4 lang3 library for using StrSubstitutor for email templates

Signed-off-by: Rohit Yadav <rohit.yadav@shapeblue.com>

yadvr committed:
quota: assume start/end date in the balance API will be returned

enhances UI, to reflect balance api changes

Signed-off-by: Rohit Yadav <rohit.yadav@shapeblue.com>

abhinandanprateek committed:
quota: adding config initialization to quota manager and taking care of boundary conditions

abhinandanprateek committed:
quota: make tariff plans updatable and schedulable

abhinandanprateek committed:
quotq: optimizing tariff retrival while calculating quota usage.

abhinandanprateek committed:
quota: fix tariff type

yadvr committed:
ui: add effective from date in tariff list/table, add error handling dialogs

Signed-off-by: Rohit Yadav <rohit.yadav@shapeblue.com>

yadvr committed:
quota: fix schema, set all quota tariff values to 0; effective date as cloud.com product launch date

Signed-off-by: Rohit Yadav <rohit.yadav@shapeblue.com>

yadvr committed:
quota: fix tariff update cmd to accept start date, update effective/start date in db

Signed-off-by: Rohit Yadav <rohit.yadav@shapeblue.com>

yadvr committed:
quota: fix db quota usage

Signed-off-by: Rohit Yadav <rohit.yadav@shapeblue.com>

yadvr committed:
quota: fix stuff

Signed-off-by: Rohit Yadav <rohit.yadav@shapeblue.com>

abhinandanprateek committed:
quota: adding final where ever required

abhinandanprateek committed:
quota: transaction commit when a account is processed

abhinandanprateek committed:
quota: renaming DBUtils to ResponseBuilder as that is what it is

abhinandanprateek committed:
quota: restructuring

yadvr committed:
quota: email template backend and apis

Signed-off-by: Rohit Yadav <rohit.yadav@shapeblue.com>

yadvr committed:
quota: escape javascript, but not html

people might want to send html emails

Signed-off-by: Rohit Yadav <rohit.yadav@shapeblue.com>

yadvr committed:
quota: add email template UI

Signed-off-by: Rohit Yadav <rohit.yadav@shapeblue.com>

yadvr committed:
quota: fix UI's escaping quotes

Signed-off-by: Rohit Yadav <rohit.yadav@shapeblue.com>

yadvr committed:
quota: fix getusagerecords name to reflect its for getting unprocessed quota cloud_usage data

Signed-off-by: Rohit Yadav <rohit.yadav@shapeblue.com>

yadvr committed:
quota: implement quota email alert methods/classes and template engine to substitute variable options

Signed-off-by: Rohit Yadav <rohit.yadav@shapeblue.com>

yadvr committed:
quota: remove include column from tariff, fix list tariff api to accept start date

Signed-off-by: Rohit Yadav <rohit.yadav@shapeblue.com>

yadvr committed:
quota: fix ui - statement, tariff and templates

Signed-off-by: Rohit Yadav <rohit.yadav@shapeblue.com>

yadvr committed:
ui: fix UI, API usage and display rendering; fix template saving to use POST

Signed-off-by: Rohit Yadav <rohit.yadav@shapeblue.com>

yadvr committed:
quota: fix template update API to increase body/subject size/length

Signed-off-by: Rohit Yadav <rohit.yadav@shapeblue.com>

yadvr committed:
quota: fix balance api to send result when no debits have been made

Signed-off-by: Rohit Yadav <rohit.yadav@shapeblue.com>

yadvr committed:
quota: add logic to send email using a deferred approach

a deferred bg thread could be later used to send emails

Signed-off-by: Rohit Yadav <rohit.yadav@shapeblue.com>

abhinandanprateek committed:
quota: statement genertaion date should not be in future

yadvr committed:
quota: fix stuff

Signed-off-by: Rohit Yadav <rohit.yadav@shapeblue.com>

yadvr committed:
ui: show actual backend error, if it's not auth failure

Signed-off-by: Rohit Yadav <rohit.yadav@shapeblue.com>

yadvr committed:
quota: remove unused import

Signed-off-by: Rohit Yadav <rohit.yadav@shapeblue.com>

abhinandanprateek committed:
quota: account for 0 tariff

abhinandanprateek committed:
Merge branch 'quota' of https://github.com/shapeblue/cloudstack into quota

abhinandanprateek committed:
quota: added a method to get the last balance in balancedao

abhinandanprateek committed:
quota: rename method to lastQuotaBalance

yadvr committed:
quota: only lock normal accounts

Signed-off-by: Rohit Yadav <rohit.yadav@shapeblue.com>

yadvr committed:
quota: on getting enough credits, unlock the account which got new credits

Signed-off-by: Rohit Yadav <rohit.yadav@shapeblue.com>

yadvr committed:
quota: for checking/sending email use the new lastQuotaBalance method to get latest balance

Signed-off-by: Rohit Yadav <rohit.yadav@shapeblue.com>

abhinandanprateek committed:
quota: get last balance api

yadvr committed:
quota ui: grab current date in the browser's tariff

Signed-off-by: Rohit Yadav <rohit.yadav@shapeblue.com>

yadvr committed:
quota: unlock/lock logic with addCredits api

Signed-off-by: Rohit Yadav <rohit.yadav@shapeblue.com>

abhinandanprateek committed:
quota: only add the initial credit entries to the balance as they have
not yet been accounted for

yadvr committed:
quota: don't lock account, only unlock with adding credits

Signed-off-by: Rohit Yadav <rohit.yadav@shapeblue.com>

abhinandanprateek committed:
quota: allow locked accounts to be credited

yadvr committed:
quota: add new template for sending unlocked email account

yadvr committed:
quota: add currency to list/readonly apis

Signed-off-by: Rohit Yadav <rohit.yadav@shapeblue.com>

yadvr committed:
quota: fix ui to grab currency from the response

Signed-off-by: Rohit Yadav <rohit.yadav@shapeblue.com>

yadvr committed:
quota: add quota monthly statement email template

Signed-off-by: Rohit Yadav <rohit.yadav@shapeblue.com>

abhinandanprateek committed:
quota fixed misc boundary conditions

abhinandanprateek committed:
quota: fixed permisison and date flip

yadvr committed:
quota: add currency in case of users too

Signed-off-by: Rohit Yadav <rohit.yadav@shapeblue.com>

yadvr committed:
quota: send currency with add credits response

Signed-off-by: Rohit Yadav <rohit.yadav@shapeblue.com>

yadvr committed:
quota: fix css of the ui, make element similar to other cloudstack UI elements

Signed-off-by: Rohit Yadav <rohit.yadav@shapeblue.com>

abhinandanprateek committed:
quota: restructured/refactored to make quota run inside the usage server

abhinandanprateek committed:
quota: run the calculations after the usage server is done collecting
usage

yadvr committed:
quota: improve css and element usage; align inputs below labels

Signed-off-by: Rohit Yadav <rohit.yadav@shapeblue.com>

abhinandanprateek committed:
quota: adding book keeping columns in accounts table

abhinandanprateek committed:
quota: fine tuning dates

yadvr committed:
quota: fix stuff

Signed-off-by: Rohit Yadav <rohit.yadav@shapeblue.com>

yadvr committed:
quota/usage: fix license

Signed-off-by: Rohit Yadav <rohit.yadav@shapeblue.com>

abhinandanprateek committed:
quota: added quota_account

abhinandanprateek committed:
quota: fixing account bean

abhinandanprateek committed:
quota: tariff fix

abhinandanprateek committed:
quota: fixing warped up db

abhinandanprateek committed:
quota: added min_balance and quota enforce for each account

yadvr committed:
quota ui: add min threshold and checkbox to enforce locking

Signed-off-by: Rohit Yadav <rohit.yadav@shapeblue.com>

abhinandanprateek committed:
quota: due to interpackage dependency the account lock functionality is pulled
into usage

abhinandanprateek committed:
quota: even if there is no usage for some resources show them in the
statement.

abhinandanprateek committed:
quota: fixes

yadvr committed:
quota: minor credits ui fixes

Signed-off-by: Rohit Yadav <rohit.yadav@shapeblue.com>

yadvr committed:
quota: fix database

Signed-off-by: Rohit Yadav <rohit.yadav@shapeblue.com>

abhinandanprateek committed:
quota: refactored email code and pulled it into ALertManager that will
run after Quta calc's are done

abhinandanprateek committed:
quota: added checks for quota enforce and threshold per account

yadvr committed:
quota: show ui balance if start/end balance are defined (any values)

Signed-off-by: Rohit Yadav <rohit.yadav@shapeblue.com>

abhinandanprateek committed:
quota:fixes

abhinandanprateek committed:
Revert "quota: fixes"

This reverts commit fc95ece6cbc4f9d6e633e025b8efb97525175783.

abhinandanprateek committed:
quota: streamlined emailing

abhinandanprateek committed:
quota: +send last month statement at the start of the month

abhinandanprateek committed:
quota: code cleanup

yadvr committed:
quota: check if quota is enabled, only then expose APIs

Signed-off-by: Rohit Yadav <rohit.yadav@shapeblue.com>

yadvr committed:
quota;e2etesting

Signed-off-by: Rohit Yadav <rohit.yadav@shapeblue.com>

abhinandanprateek committed:
quota: though default value for quota_calculated is 0, the usage server
makes it null while entering usage entries. Flipping the condition so
as to acocunt for that.

yadvr committed:
quota: remove quotaRefresh

Signed-off-by: Rohit Yadav <rohit.yadav@shapeblue.com>

abhinandanprateek committed:
quota: by default make the global quota enforement flag as false. fix
the condition so that the uasage generated records are picked up in
quota calc cycle

abhinandanprateek committed:
quota: to maintain idempotency the alter table statement is moved to
java upgrade code. In case the column already exists the code will
ignore the error.
